### ISSUE TITLE
VoiceOver support on iPad

### ISSUE THREAD

**Issue by giulianozor** [isontheline/pro.webssh.net#1167]
# Feature description

thanks

**Comment by isontheline** [comment]
@giulianozor I need more information please

**Comment by giulianozor** [comment]
@isontheline 
Hello, the main issues I face are;
– When using the keyboard, I get the voice feedback when inserting text, but not when deleting text.
– I cannot get the terminal content reading. The only way is to enable screen recognition, but this will read everything from the beginning, so it is not very useful. The best would be to have each line selectable and readable.
– The Corso to move up and down, left, and right in the upper row of the keyboard is not usable when VoiceOver is enabled.
That is all I can think of right now.

Thank you.
G
 G

**Comment by isontheline** [comment]
Thanks for your feedback @giulianozor !

> – When using the keyboard, I get the voice feedback when inserting text, but not when deleting text.

I'm new to VoiceOver so I need some guidance.
When I delete chars I hear a sound, is it not the intended behaviour?

> – I cannot get the terminal content reading. The only way is to enable screen recognition, but this will read everything from the beginning, so it is not very useful. The best would be to have each line selectable and readable.

I'm not able to provide text selection on the terminal due to an underlying library, you need to use "Show Scrollback Buffer"
I know that's not ideal so I will try to make a workaround

> – The Corso to move up and down, left, and right in the upper row of the keyboard is not usable when VoiceOver is enabled.

Are you talking about this component?
<img width="435" alt="image" src="https://github.com/user-attachments/assets/4f89d77d-9998-4d45-b784-3ac6d1f1a009">

**Comment by giulianozor** [comment]
@isontheline
When VoiceOver is enabled you get the reading of each character you type or delete. For example, if you type 123, you can hear one, two, three. Then if you delete 32, you hear three, two. This allows you to know which character you typed or delete. Also, there are other mouths like words or lines where you get the reading for a whole for the line, but in this case only when you delete, move or select a different line or worth.
If it is not possible to get the liner readings without entering the buffet screen, that is very unfortunate Punto it makes the whole software 99% useless because there is no easy way to receive the output of your comments. Maybe the underlying library could be updated to support voiceover? I do not use it on Mac, only on iPad, but I suppose the same issue is present in the desktop version as well.
The comp you circled in the AMEX is exactly the one I am talking about.

Thanks.


G

PS: sorry for not quoting correctly, it is difficult for me to quote using dictation

**Comment by isontheline** [comment]
> PS: sorry for not quoting correctly, it is difficult for me to quote using dictation

No problem at all

> that is very unfortunate Punto it makes the whole software 99% useless because there is no easy way to receive the output of your comments

I will improve WebSSH in the upcoming releases to have a better accessibility

If you are still using WebSSH, I'll ping you when I've made some improvements to the accessibility elements

**Comment by giulianozor** [comment]
@isontheline 
Yes, I still use it, at least while I have some vision left. I will be glad to provide feedback and test.

I also tried other clients, but I have not found anything even close to web SSX, so I really hope you can improve the accessibility.

Thanks

G i

**Comment by isontheline** [comment]
Hello @giulianozor 👋

WebSSH 28.1.1438 beta release is available with better VoiceOver support : https://testflight.apple.com/join/QSrBK59z

**Comment by giulianozor** [comment]
@isontheline hello,
Thank you very much for the update!
The cursor keys now working properly, and the voice reading for command outputs is great!

If possible, it would be great if you can change the voice output for the left and right arrow keys so the current character is red instead of left and right. These may or may be not related to the issue where the deleted characters are not spoken when deleted.
Also, it would be awesome if when using up and down you get the command reading, if any is retrieved from the history.

New issue: it seems it is not possible to y scroll up and down to see the screen buffer. this is usually done with a swipe up or down with three fingers.

again, thanks a lot for the update!

**Comment by giulianozor** [comment]
@isontheline hello
I noticed that the issue about spoken deleted characters is fixed if I start a session  with VoiceOver enabled, but if I start VoiceOver after connecting to a session, it is still present. In this case, the cursor key  reverts back to the original button too.  
I think that maybe an option in the settings to enable voiceover integration could work better, at least for this. I have seen this approach in other southwest, like IntelliJ idea. On the first start with VoiceOver enabled you get a pop-up that explains how to turn on the full integration in the settings. If you do not enable this option, you still get some integration, but with some limitations. if you enable this option, the changes are kept even if you turn off voiceover i.e. the keyboard shortcuts are different

Thank you.

G

**Comment by isontheline** [comment]
Hello @giulianozor 👋

Many thanks for your strong feedback 🙏

No problem I will make these improvements over the upcoming weeks and I will ping you each time a new enhancement is available.

Thank you to help me to enhance WebSSH accessibility!

**Comment by giulianozor** [comment]
@isontheline hello,
Another useful addition could be to speak the auto completion text when you press the tab key.
Thanks
G

**Comment by isontheline** [comment]
Hello @giulianozor 👋

No problem I'm planning it at same time with "When using up and down, speak the retrieved command from history"

**Comment by isontheline** [comment]
Hello @giulianozor 👋

Could you please try the latest beta release?
Specifically "Ability to scroll up / down the buffer while VoiceOver is enabled"

Thank you!

**Comment by giulianozor** [comment]
@isontheline  hello,
Thanks for the update. The school up and down seems to work correctly. In fact I prefer this way than the non-voiceover one 👍 

thanks again.
G

**Comment by isontheline** [comment]
Hello @giulianozor 👋

> Change the voice output for the left and right arrow keys so the current character is read instead of left and right.

By current, you mean character behind the blinking cursor to be read?

**Comment by giulianozor** [comment]
I mean the char with th cursor under it after the cursor is moved. Let’s say i have the word 123456 and the cursor is under the 5. If i go lett (or delete) i hear “4”m i, i go right i hear “6”

**Comment by isontheline** [comment]
Thanks for clarification @giulianozor 🙏
I will try something about it
